### DISCUSSION TITLE
this바인딩된 이벤트 리스너를 제거하는 좋은 방법은 무엇일까요?

### DISCUSSION THREAD

**Discussion by wonsss** [woowacourse-study/2022-book-muscle#50] [온스타]
온스타와 [다른 디스커션](https://github.com/woowacourse-study/2022-book-muscle/discussions/23)에서  this관련하여 이벤트 리스너가 binding되는 방식(bind, 화살표함수) 등의 방식을 논의하였는데, 그러다가 나중에 이벤트 리스너를 제거하는 것이 중요하다는 결론에 이르렀습니다.

따라서  `this바인딩되어 이벤트 리스너(콜백함수)를 제거하는 좋은 방법`에 대해 고민해보고 검색도 해보았는데요.

그 중 찾은 방법은 this 바인딩된 이벤트리스너를 constructor 안에서 멤버변수로 할당한 후, 이벤트를 추가하거나 제거하는 경우 이벤트리스너 인자 자리에 멤버변수를 넣어주는 것입니다. 

```jsx
class Marco {
  constructor() {
     this.myListener = this.myListener.bind(this);
  }
  $myElement.addEventListener('click', this.myListener);
  $myElement.removeEventListener('click', this.myListener);
}
```

이 방법에 대해서 어떻게 생각하시나요? 이 방법이 가능한지, 또는 더 적절한 방법이 있는지 의견을 듣고 싶어요.

> 참고
[SOF - removing-event-listener-which-was-added-with-bind](https://stackoverflow.com/questions/11565471/removing-event-listener-which-was-added-with-bind)

**Comment by onschan** [comment]
이 디스커션이 올라왔던 시점도 정확히 기억하는데 페어프로그래밍때문에 정신이 없어서 지금에야 댓글 다네요! 미안해요 마르코😣

[로또 미션 - 안 PR 에 대한 로이의 피드백](https://github.com/woowacourse/javascript-lotto/pull/153#discussion_r820128873)
```js
constructor() {
  ...
  $('...').addEventListener('submit', this.#method);
}
#method = (e) => { ... }
```
> this.#method.bind(this) 는 this를 bind한 새로운 함수를 생성하는 명령입니다.
어차피 새로운 함수를 만드는건 똑같으니, 이보다는 처음부터 인스턴스 자신의 메소드로 들고 있게끔, class field로 선언해주면 더 좋다고 생각합니다.

로이의 방식은 그냥 메서드로 선언하고 핸들러로 사용하는 것이아니라 클래스 필드에 익명함수를 할당하고 해당 필드를 핸들러로써 사용하는데요

메서드를 선언하고 bind(this)로 this를 바인딩하는 경우는, 이미 클래스의 프로토타입에 원본 메서드가 존재하는 상황에서 인스턴스 내부에서 새로운 함수를 다시 한번 더 생성하는 것이나 다를게 없으니 애초에 프로토타입과 관련없이 인스턴스 자체에서만 사용되는 함수를 만들어서 핸들러로 사용한다! 라는 접근 같아요. 

위와 같은 경우는 #method 에 할당된 함수가 항상 같은 주소값을 가리키니 removeListener를 해줄때도 아무런 문제가 없겠죠?

추가적으로 화살표 함수를 사용한 이유는 this가 가리키는 대상에 대한 모호함을 상쇄하기 위함인거 같기도 해요.

저도 겉핡기 식으로 이해한 부분이라 이 부분에 대해서 마르코의 의견 혹은 다른 분들의 의견도 궁금하네요!

**Reply by wonsss** [reply]
온스타. 덕분에 잘 이해했습니다. 감사합니다.

bind를 하는 것이나 화살표 함수로 메서드를 만드는 것이나, 모두 새 함수를 만든다는 점에서는 것은 똑같으니 `성능`상 차이는 없겠군요!

다만, 로이가 알려준 것처럼 `화살표 함수 메서드는 인스턴스의 메서드`가 될 수 있고, bind는 그렇지 못하고 아마 별개의 영역(bound?)에 선언되기 때문에, 객체 지향답게 인스턴스 내부에 선언되는 화살표함수 메서드가 더 나을 수 있다는 거겠네요.

제가 이해한 내용이 맞을까요?

그리고 애매한 부분은 bind한 함수가 만들어지는 곳인데요. 그 곳은 어디이고 부르는 정확한 명칭이 있을까요? 😄
